Doodle

romanpeters opened the pull request:
### Keyword functionality
Keywords can be used to trigger functions, just like commands can.
The main differences are:
- No special character needed to trigger
- Keywords can be anywhere within the message text, even within words in the text

Since keywords triggering things can be very annoying, this feature should generally not be used. Use commands instead.

### New database table
The table `database.Doodle` can be used to persistently store Doodle polls.

### Doodle Saving
When a URL to a Doodle poll is posted (keyword: "https://doodle.com/poll/"), this URL is saved to the database.

### Doodle Retrieving
The command `!doodle` returns the last Doodle poll URL saved to the database.

romanpeters committed:
Add db table for states in the form of strings

romanpeters committed:
Add @Keyword registry

romanpeters committed:
Save and retrieve Doodle URL's

romanpeters committed:
Version increase

romanpeters committed:
Trigger @Keyword with message content

romanpeters committed:
Doodle not found message

romanpeters committed:
Remove trailing whitespace

romanpeters commented:
Python Black keeps crashing on me, so if someone else could Black this before merging, that would be great.

RobinSikkens commented on devbot/database.py:
```diff
@@ -51,4 +51,11 @@ class XKCD(SQLAlchemyBase):
 
 
+class StringStates(SQLAlchemyBase):
+    """ Variables for the bot to keep track of """
```
Please add `.` to your docstrings

RobinSikkens commented on devbot/commands/doodle.py:
```diff
@@ -0,0 +1,27 @@
+
+@Command(["doodle"])
+async def doodle(*_args, **_kwargs):
+    """ Return the last registered Doodle URL """
```
Please add `.` to your docstring

RobinSikkens commented on devbot/commands/doodle.py:
```diff
@@ -0,0 +1,27 @@
+
+@Keyword(["https://doodle.com/poll/"])
+async def doodle_url(message_contents, *_args, **_kwargs):
+    """ Called when a message contains a Doodle poll URL and saves it to the database """
```
Please add `.` to your docstring

RobinSikkens commented on devbot/registry.py:
```diff
@@ -72,6 +74,23 @@ def __init__(self, command):
+    target_dict = KEYWORD_DICT
+
+    def __init__(self, name):
+        """ Store params and call super """
```
Please add `.` to your docstrings

RobinSikkens commented on devbot/bot.py:
```diff
@@ -55,10 +56,21 @@
+            response = await safe_call(
+                KEYWORD_DICT, keyword, message_contents, message, CLIENT
+            )
+            LOGGER.info("keyword")
```
Please elaborate in this info message, also should this INFO or DEBUG?

romanpeters replied:
It mirrors [line 54](https://github.com/DevOps-Utrecht/bot/blob/f409e1c8a8af1321286469e20bf61e9fd698b4fc/devbot/bot.py#L54). Should I convert both to DEBUG?

RobinSikkens replied:
It's okay for it to be info if you could change them both to be more descriptive however that would be great.

romanpeters committed:
Docstring now has her period

romanpeters committed:
More descriptive logging

RobinSikkens committed:
Blackify everything

romanpeters committed:
Better database architecture

romanpeters committed:
Merge branch 'doodle' of https://github.com/DevOps-Utrecht/bot into doodle

romanpeters committed:
Revert "Merge branch 'doodle' of https://github.com/DevOps-Utrecht/bot into doodle"

This reverts commit 5e034e81d1c4f312c948d1aa8342af517aad28bc, reversing
changes made to f24f7014e060853699cba518bdccd554522307ff.

romanpeters committed:
Bug fixes

romanpeters committed:
Black once again
for the renegade master
D4 damager, power to the people
Back once again for the renegade master
D4 damager, with the ill behavior

RobinSikkens commented on devbot/bot.py:
```diff
@@ -55,10 +56,21 @@
+            response = await safe_call(
+                KEYWORD_DICT, keyword, message_contents, message, CLIENT
+            )
+            LOGGER.info("keyword")
```
```suggestion
            LOGGER.info(f"Keyword: {keyword} was triggered.")
```

romanpeters replied:
Huh, I thought I had fixed that

RobinSikkens commented on devbot/database.py:
```diff
@@ -51,4 +51,13 @@ class XKCD(SQLAlchemyBase):
+    """ Representation of a Doodle poll. """
+
+    __tablename__ = "doodle"
+    id = sa.Column(sa.Integer, primary_key=True, autoincrement=True)
```
Using id shadows a function, this might cause issues. Check the [docs](https://docs.python.org/3/library/functions.html#id) for more info.

romanpeters replied:
Good point

romanpeters committed:
Better logging

romanpeters committed:
database.Doodle.id changed to database.Doodle.num

romanpeters committed:
Iterating the dictionary directly instead of calling .keys()

romanpeters committed:
Prevent redefining name 'doodle_url' from outer scope

romanpeters committed:
Remove useless super delegation

romanpeters committed:
Remove unnecessary "else" after "return"

romanpeters committed:
Merge branch 'master' into doodle

romanpeters commented:
I can't merge it without a new approving review, but feel free to do so. Should be all good.
